### PR TITLE
Implement parsing through enums/structs instead of vecs/strings

### LINKED ISSUE #134: Use Enums for the AST

### PR DESCRIPTION
**Pull request by itsjunetime**
... and adjust rest of workspace to work with it.

Closes #134.

Sorry that this is all one commit, I couldn't think of a way to split it up since basically every part of this depends on another part.

The big change comes in the `rs/parser/src/nodes.rs` file - this implements all the various structs and enums that exist in the grammar, as best as I could determine. When I was converting the main crate over to parsing this AST, I ran into a few situations where the grammar in `glicol.pest` and the way that it was turned into `glicol_synth` nodes seemed to disagree, and I just had to make my best judgement in those situations. Sometimes I modified the struct in the AST, and sometimes I just added a `panic!` with an explicit error message about how that feature/syntax isn't supported yet. If the `panic` message is inaccurate and that feature won't ever be supported, let me know and I'll modify the structs to fail at parsing that sort of thing.

The second biggest changes are in `rs/main/src/{lib,util}.rs`. Two things happen here:
1. In `util.rs`, the AST is converted into nodes to be placed on the graph. This was fairly straightforward, and allowed us to remove a ton of `unimplemented!` calls (since all enums used are now exhaustive and have no invalid states, instead of just using `GlicolPara` as a catch-all).
2. In `lib.rs`, I changed the way we update the engine with code. Previously, we would update the code, then set the flag `needs_update`, then actually diff and update the graph the next time `next_block` was called. I felt like this was not ideal, as it causes the AST and code to be out-of-sync. So instead I switched it to make `update_with_code` actually do the diffing and updating of the syntax tree, and `next_block` just gets the next block of sound.
    This required some undesirable combination of methods into one large method that does a bunch of stuff, but I think that it's for the best. It allows us to remove a bunch of members from the `Engine` struct that are only there temporarily for diffing, and put those into another struct that borrows from `self` and the new `YokedAst` for diffing, instead of copying data over to the temporary members of `self` during the diffing process.
    I also got rid of the `lcs_diff::diff` method and elected to do diffing manually, as that method required cloning of each AST struct that was passed in, and that's an unnecessary cost to pay. Instead, we just iterate through each new and old node + chain and see what stayed the same, what was removed, and what was added.

A few other minor things:
- `GlicolPara` is now generic over `S` so that we can use it in situations where we want to borrow from something else to create like a `GlicolPara::Reference` to avoid another allocation
- The code that writes an `EngineError` to a `&mut [u8]` is now moved to the `wasm` crate, since that's the only thing that actually needs that code.
- The `wasm` crate is now part of the workspace since it needs to access `pest` (to match details of the `EngineError`, so its profile details have been moved to a profile called `wasm-release` (instead of just using the `release` profile, as was previously done)
- `pest` errors are now boxed in our return types since they are quite large and thus make every `Result` type that contains them quite large as well, incurring a significant cost on method returns even if we don't encounter an error. Now, they shouldn't be quite as large, at the cost of only one allocation *if* we run into an error (as opposed to incurring costs before regardless of if we ran into an error).

The main issue with this all is that I haven't actually tested it yet. Before merging, I'd like to get some more formal tests written inside the parser crate, and I'd also like to just try this out with something like `glicol-cli` to make sure everything seems to still work just fine.

On that note, this should also allow much clearer error messages in `glicol-cli` that can point to exactly where we failed to parse. This should make for a much better experience there, once this gets merged and I can push a PR to `glicol-cli` to work with this new API.

Of course, this will also change the public API, so all these crates should be bumped to `0.14.0` if this gets merged.

If you have any questions, obviously let me know! And please don't merge until I've got some tests in (just as some motivation for myself).

### CODE REVIEW

**Comment by itsjunetime** [comment]
I've ran this with some of my own code and it looks like the parsing works correctly now (though I'd still like to write some more tests for it), but the diffing has issues (it panicked upon a code update being ran with glicol-cli) so I need to fix that.

**Comment by itsjunetime** [comment]
Ok, I think I may have fixed the panicking thing (i forgot a single `!`) - there are still more tests I want to add, but I think the main part of the code is reviewable now.

**Comment by chaosprint** [comment]
I just tested the WASM part a little bit: when there is an syntax error, we get a panic and the sound stops. Ideally, the music should continue and only report the error.

**Comment by itsjunetime** [comment]
Would you be able to share how you're testing it in wasm, or what the stack trace is? I've only been trying this with glicol-cli so far, so I haven't had a chance to try it out with the wasm module, though I did try to make sure it still worked there.

**Comment by chaosprint** [comment]
> Would you be able to share how you're testing it in wasm, or what the stack trace is? I've only been trying this with glicol-cli so far, so I haven't had a chance to try it out with the wasm module, though I did try to make sure it still worked there.

First run `build.sh` in the ./rs/wasm folder. But since the `wasm` folder is now part of the workspace, we need to accordingly change the `./target` folder path, which I have done in this branch:

https://github.com/chaosprint/glicol/tree/new-parser-js-test

Diffs can be found here: https://github.com/chaosprint/glicol/commit/351605aa5ca3f214e773d8f21c7636a0317e009c

Next, in the ./js folder, run `pnpm i` and then `pnpm dev --open` (you need to have [pnpm](https://pnpm.io/) installed).

In the browser (preferably Chrome) we can see a simple editor for testing and all the error message from the browser console.

Let me know if you have any questions. But in generally, I think it's fine to leave the wasm/js part broken and solve it later since this is a breaking change targeting 0.14 anyway. I will test glicol-cli as well tmr.

**Comment by itsjunetime** [comment]
Ok, I finally tried out the js build and it seems to work fine on my machine - the pretty-print errors even appear in the console now, which seems cool.

As one side note - it might be smart to run `wasm-opt` (from [binaryen](https://github.com/WebAssembly/binaryen)) on the final wasm bundle before uploading to npm. It took it from ~3.4M to 2.3M on my machine, which would probably be nice for end users.

Also, would you (@chaosprint) be able to look at the `TODO`s that I have in the code in this PR and see what ones you could answer/fix? Some of them regard the naming of fields in structs, and some others are about specific nodes for which I couldn't find example syntax (and thus couldn't write test cases for). I'm also wondering about some syntax ambiguities within the `seq` node and hoping to hear your thoughts about those ambiguities before trying to fix the failing tests there.

**Comment by chaosprint** [comment]
Sorry I somehow missed this update... I will test it ASAP...

**Comment by chaosprint** [comment]
> Ok, I finally tried out the js build and it seems to work fine on my machine - the pretty-print errors even appear in the console now, which seems cool.
> 
> As one side note - it might be smart to run `wasm-opt` (from [binaryen](https://github.com/WebAssembly/binaryen)) on the final wasm bundle before uploading to npm. It took it from ~3.4M to 2.3M on my machine, which would probably be nice for end users.
> 
> Also, would you (@chaosprint) be able to look at the `TODO`s that I have in the code in this PR and see what ones you could answer/fix? Some of them regard the naming of fields in structs, and some others are about specific nodes for which I couldn't find example syntax (and thus couldn't write test cases for). I'm also wondering about some syntax ambiguities within the `seq` node and hoping to hear your thoughts about those ambiguities before trying to fix the failing tests there.

opps I forgot to mention that you need to comment out the version number to test locally here:
image


when I test it today, it cannot report the error in console. instead just panic info and the sound stops. 


the original wasm/src/lib.rs is working with the `js/src/glicol-engine.js`, especially the `process` func.

the Rust side should always yield audio samples. if there is a grammar mistake, report it through the result ptr but continue to yield samples using previously correct grammar input.

could you pls fix this?

**Comment by itsjunetime** [comment]
Alright, it took a second but I've got it working now with the website. I haven't updated the scripts under `js/npm` 'cause they seemed slightly different from everything under `js/src` in ways that I couldn't figure out easily, but audio should now continue playing when you have a syntax error and all syntax errors should be reported in the console.

Unfortunately, to get all these changes in easily, I did have to switch to using wasm-bindgen (I needed extra debugging tools to actually figure out what was going wrong, and they all required use of wasm-bindgen to work correctly), but I think that's a good move in the long run. wasm function definitions are now generated for us, so js lsps can check function parameters and all that kinda stuff, and this also allowed us to get rid of a lot of now-unused code (such as the RingBuffer and TextParameterWriter)

Let me know if it looks good and works for you. The `build.sh` file is also updated for building with wasm-bindgen, so you should be able to have the same development flow as before.

**Comment by chaosprint** [comment]
> Alright, it took a second but I've got it working now with the website. I haven't updated the scripts under `js/npm` 'cause they seemed slightly different from everything under `js/src` in ways that I couldn't figure out easily, but audio should now continue playing when you have a syntax error and all syntax errors should be reported in the console.
> 
> Unfortunately, to get all these changes in easily, I did have to switch to using wasm-bindgen (I needed extra debugging tools to actually figure out what was going wrong, and they all required use of wasm-bindgen to work correctly), but I think that's a good move in the long run. wasm function definitions are now generated for us, so js lsps can check function parameters and all that kinda stuff, and this also allowed us to get rid of a lot of now-unused code (such as the RingBuffer and TextParameterWriter)
> 
> Let me know if it looks good and works for you. The `build.sh` file is also updated for building with wasm-bindgen, so you should be able to have the same development flow as before.

the js part is working now! great. I agree that using wasmbindgen makes the code much cleaner and easier to read.

the only issue I found for now is the performance seems to be degraded heavily after switching to the wasm bindgen.

my intuition is that lots of GC is happening in the audioworklet with the binding js code.

this makes the `chaos` example almost impossible to run and the rest examples are at the edge of underrun.

previous solution is kinda "unsafe" and requires some manual checking but it's more efficient in terms of audio performance.  

https://github.com/padenot/ringbuf.js is adopted in the previous solution to avoid GC when passing params or code. It has to be defined twice in the main thread and the audioworklet thread. yeah it's not the most ergonimic..

the only thing we need to fix is the error msg passing to JS duing code update. I suggest that we ignore this in this pr and fix this later. I plan to seperate create a new repo called glicol.js to focus on the CDN and NPM package, which will make the CI much easier to setup later.

**Review by chaosprint — CHANGES_REQUESTED** [review]
just need to change the `sp` test

**Comment by itsjunetime** [comment]
I just pushed some changes to (hopefully) fix the performance issues. Previously, we would just pass the code in as a js string to the audio worklet thread, which was (in that thread) copied, encoded to utf8, and then copied again to pass it to rust as a string. Now, we encode it to utf-8 outside of that thread and then pass it into the worklet thread, where it is handed off to rust to be (unsafely) interpreted as a valid utf8 string. The chaos demo works perfectly fine for me with these changes; let me know if it works for you too.

**Comment by chaosprint** [comment]
> I just pushed some changes to (hopefully) fix the performance issues. Previously, we would just pass the code in as a js string to the audio worklet thread, which was (in that thread) copied, encoded to utf8, and then copied again to pass it to rust as a string. Now, we encode it to utf-8 outside of that thread and then pass it into the worklet thread, where it is handed off to rust to be (unsafely) interpreted as a valid utf8 string. The chaos demo works perfectly fine for me with these changes; let me know if it works for you too.

This doesn't compile. Something to do with ConstSig and Signal, can you fix that? I managed to use your #70bc44e 's definition of ConstSig and got it to compile and test, and the performance is great.

**Comment by itsjunetime** [comment]
> This doesn't compile

Ah, yeah sorry about that - should be fixed now

> So this would be the last tweak I believe

Sorry, not quite - there are still a few more TODOs in the `parser/src/nodes.rs` file that I was hoping you'd be able to comment on since I wasn't able to figure out what do to there. Specifically the `seq` one - some tests are failing due (what I think are) the parsing ambiguities in the `seq` grammar, so we need to resolve that before merging this.

File: rs/parser/src/nodes.rs [hunk 1]
```diff
@@ -0,0 +1,924 @@
...
+}
+
+impl Node<'_> for Plate {
+    #[cfg_attr(test, trace::trace(prefix_enter = "[+ Plate]"))]
+    fn parse_from_iter(pairs: &mut Pairs<'_, Rule>, span: Span<'_>) -> Result<Self, Box<Error<Rule>>> {
+        pairs.next_parsed(span).map(|mix| Self { mix })
+    }
+}
+
+// TODO: There are some ambiguous 'seq' inputs, afaict.
```
Comment (chaosprint) [inline rs/parser/src/nodes.rs:392]: try https://pest.rs/#editor and paste the glicol.pest

`seq o_` or `seq _o` should not compile.

`seq` takes ref with tilde (maybe we should call it control sig ref in the future?) such as `~a`, `60`(note) or `_` (rest).

so `seq ~o_` will play ~o at bar portion 0.0, and play a "rest" at bar portion 0.5. the latter will be automatically ignored of course. 

`seq _~o` will play ~o at 0.5.

`seq 6060` is invalid musically as 6060 will be seen as a (extremely high) midi note...

use `seq 60_60_` or simply write `seq 60 60`

same for `seq ~a_14_` 

the idea is always to divide one bar with spaces first; the divided part is called `compound note`, which is a permutation of tilde ref, note number and rest. then we further divide the compound note based on the amount of the element inside

hope this answers your question

File: rs/parser/src/nodes.rs [hunk 2]
```diff
@@ -0,0 +1,924 @@
...
+                    / elements_n as f32;
+                let e_span = element.as_end_span();
+                let e = element.into_inner()
+                    .next()
+                    .ok_or_else(|| e_span.to_err_with_positives(positives))?;
+
+                let time = relative_time_sub + relative_time_base;
+
+                match_or_return_err!(e,
+                    // TODO: We only match on integer here, but we store it as a f32. We should
```
Comment (chaosprint) [inline rs/parser/src/nodes.rs:439]: I think glicol.pest already limits it to int, with an alias `note`

File: rs/parser/src/nodes.rs [hunk 3]
```diff
@@ -0,0 +1,924 @@
...
+
+        Ok(Self { symbol, attack, decay })
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub struct PatternSynth<'ast> {
+    pub symbol: &'ast str,
+    // TODO: This isn't named in the grammar so i don't know what to call it
+    pub p2: f32
```
Comment (chaosprint) [inline rs/parser/src/nodes.rs:705]: `span`

File: rs/parser/src/nodes.rs [hunk 4]
```diff
@@ -0,0 +1,924 @@
...
+    }
+
+    // SAFETY: We must ensure that we actually wrote every item in the array, which did happen
+    // because we iterated 0..N and wrote everything
+    Ok(array.map(|t| unsafe { t.assume_init() }))
+}
+
+// TODO: What are the actual param names?
+#[derive(PartialEq, Debug)]
+pub struct Reverb {
```
Comment (chaosprint) [inline rs/parser/src/nodes.rs:864]: dampening
room_size
width
wet
dry

**Comment by itsjunetime** [comment]
Ok, I've fixed the issues I mentioned above. The last thing I'd like to do before merging this is get some tests in for parsing `point` node grammar, but I'm not quite sure what that would look like, so I'm ok with ignoring that in the interest of getting this merged.

**Review by chaosprint — APPROVED** [review]

**Pull request merged**